pkg/controller/tektoninstallation/tektoninstallation_controller.go

```diff
@@ -0,0 +1,167 @@
+)
+
+var log = logf.Log.WithName("controller_tektoninstallation")
+
+// Add creates a new TektonInstallation Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	cr := toolchain.NewTektonInstallation()
+	if err := mgr.GetClient().Create(context.TODO(), cr); err != nil && !errors.IsAlreadyExists(err) {
+		return errs.Wrapf(err, "failed to create tekton installation custom resource")
+	}
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileTektonInstallation{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("tektoninstallation-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource TektonInstallation
+	if err := c.Watch(&source.Kind{Type: &v1alpha1.TektonInstallation{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{}); err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource
+	enqueueRequestForOwner := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &toolchainv1alpha1.TektonInstallation{},
+	}
+
+	return c.Watch(&source.Kind{Type: &olmv1alpha1.Subscription{}}, enqueueRequestForOwner)
+}
+
+// blank assignment to verify that ReconcileTektonInstallation implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileTektonInstallation{}
+
+// ReconcileTektonInstallation reconciles a TektonInstallation object
+type ReconcileTektonInstallation struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a TektonInstallation object and makes changes based on the state read
+// and what is in the TektonInstallation.Spec
+func (r *ReconcileTektonInstallation) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling TektonInstallation")
+
+	// Fetch the TektonInstallation instance
+	tektonInstallation := &toolchainv1alpha1.TektonInstallation{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, maybe someone deleted it, but we need this to be exist, so creating it.
+			cr := toolchain.NewTektonInstallation()
+			if err := r.client.Create(context.TODO(), cr); err != nil {
```

Comment:
Generally we avoid creating primary_object in controller which managing the object.  Here we are creating `TektonInstallation` in it's own controller.  Hope this point is already considered.  Can you plz clarify reasoning behind it.

Reply:
That's correct @nurali-techie, but our requirement here is to create required customresources like   `CheInstallation/TektonInstallation` automatically by our operator, we don't want user to create this custom resources, as this is simplified opinionated flow for user. This is why we are created it upfront before starting controller and also in reconcile to handle use case if user accidentally deletes `TektonInstallation`

Reply:
@dipak-pawar thanks for clarification.